drivers/sensor/tmag5273/tmag5273_trigger.c

```diff
@@ -0,0 +1,444 @@
+   }
+
+   int retval = 0;
+
+#ifdef CONFIG_PM_DEVICE
+   retval = tmag5273_pm_is_modifiable(dev);
+
+   if (0 != retval) {
+      return retval;
+   }
+
+#endif
+
+   const struct tmag5273_config* drv_cfg = dev->config;
+   struct tmag5273_data* drv_data = dev->data;
+
+   // ---
+
+   retval = tmag5273_trigger_mask_interrupt(dev, true);
+
+   if (0 > retval) {
+      return retval;
+   }
+
+   if (SENSOR_CHAN_MAGN_XYZ == trig->chan) {
+      uint8_t value = 0;
+
+      switch (trig->type) {
+         case SENSOR_TRIG_DATA_READY: {
+            value = TMAG5273_INT_RSLT_INT_ENABLED << TMAG5273_INT_RSLT_INT_EN_POS;
+            break;
+         }
+
+         case SENSOR_TRIG_THRESHOLD: {
+            value = TMAG5273_INT_THRSLD_INT_ENABLED << TMAG5273_INT_THRSLD_INT_EN_POS;
+            break;
+         };
+
+         default: {
+            LOG_DBG("unknown trigger type |%d|", trig->type);
+            retval = -ENOTSUP;
+
+            break;
+         }
+      }
+
+      if (0 == retval) {
+         drv_data->int_trigger = *trig;
+         drv_data->int_handler = handler;
+
+         // deactivate old interrupt and install new
+         retval = i2c_reg_update_byte_dt(&drv_cfg->i2c,
+                                         TMAG5273_REG_INT_CONFIG_1,
+                                         TMAG5273_INT_RSLT_INT_EN_MSK | TMAG5273_INT_THRSLD_INT_EN_MSK,
+                                         value);
+
+         if (0 > retval) {
+            LOG_DBG("could not enable conversion complete interrupt |%d|", retval);
+            retval = -EIO;
+         }
+      }
+   } else {
+      LOG_DBG("unsupported sensor channel |%d|", trig->chan);
+      retval = -ENOTSUP;
+   }
+
+   int retval_act = tmag5273_trigger_mask_interrupt(dev, false);
+
+   // only update if previous action was succesful
+   if (0 <= retval) {
+      retval = retval_act;
+   }
+
+   return retval;
+}
+
+int tmag5273_trigger_init(const struct device* dev)
+{
+   if (NULL == dev) {
+      LOG_DBG("dev: NULL");
+      return -EINVAL;
+   }
+
+   const struct tmag5273_config* drv_cfg = dev->config;
+   struct tmag5273_data* drv_data = dev->data;
+
+   /* set up interrupt GPIO */
+   if (!device_is_ready(drv_data->int_gpio.port)) {
+      LOG_DBG("interrupt GPIO not ready");
+      return -ENODEV;
+   }
+
+   /* configure trigger register */
+   drv_data->dev = dev;
+
+   int retval = 0;
+   uint8_t regdata = 0;
+
+   regdata = TMAG5273_SET_REGISTER_BITS(regdata, TMAG5273_INT_RSLT_INT_EN, TMAG5273_INT_RSLT_INT_DISABLED);
+   regdata = TMAG5273_SET_REGISTER_BITS(regdata, TMAG5273_INT_THRSLD_INT_EN, TMAG5273_INT_THRSLD_INT_DISABLED);
+
+   regdata = TMAG5273_SET_REGISTER_BITS_IF_CONFIG(regdata, TMAG5273_INT_STATE, TMAG5273_INT_STATE_LATCHED);
+   regdata = TMAG5273_SET_REGISTER_BITS_IF_CONFIG(regdata, TMAG5273_INT_STATE, TMAG5273_INT_STATE_PULSE);
+
+   const uint8_t int_mode =
+      COND_CODE_1(IS_ENABLED(TMAG5273_INT_MODE_INT), (TMAG5273_INT_MODE_INT), (TMAG5273_INT_MODE_NONE))
```

Comment:
I think you likely want this to be specified via Device Tree, since then it would be compile-time const but configurable on a per-device-instance basis.

Reply:
Device Tree fixes a lot of things 😁

Reply:
As far as I can see (looking at some drivers), most sensor drivers are only supporting one instance and putting all the stuff in the Kconfig. One exception I have found is the Bosch BME680, but even there the Device Tree usage is minimal. Maybe there should be more refactoring be done with new Zephyr major versions?